kbase-extension/static/kbase/js/common/simpleLogViewer.js

```diff
@@ -0,0 +1,818 @@
+                                        }),
+                                    ]
+                                ),
+                            ]
+                        ),
+                    ]
+                );
+            }
+
+            /**
+             * render the HTML structure that displays
+             * an individual job log line
+             * <ol class="${cssBaseClass}__log_line_container">
+             *    <li class="${cssBaseClass}__line_text">starting job</div>
+             *    <li class="${cssBaseClass}__line_text">initialising variables</div>
+             *    ...
+             * </ol>
+             * @param {object} line
+             */
+            renderLogLine(line) {
+                const errorSuffix = line.is_error ? '--error' : '';
+                return t('li')(
+                    {
+                        class: `${cssBaseClass}__line_text${errorSuffix}`,
+                    },
+                    line.line
+                );
+            }
+
+            /**
+             * Render the log
+             *
+             * This replaces the log wholesale with the content of model.getItem('lines')
+             * rather than keeping track of which lines have and haven't been rendered.
+             */
+            renderLog() {
+                const lines = this.model.getItem('lines');
+                if (!lines || lines.length === 0) {
+                    this.ui.setContent(LOG_PANEL, 'No log entries to show.');
+                    this.renderButtons();
+                    return;
+                }
+
+                const panel = this.getLogPanel();
+                const renderedLines = panel.querySelectorAll('ol li');
+                if (renderedLines && renderedLines.length) {
+                    const olElement = panel.querySelector('ol');
+                    // append the new lines to the existing log lines
+                    const newLines = lines
+                        .slice(renderedLines.length)
+                        .map((line) => this.renderLogLine(line))
+                        .join('\n');
+                    olElement.innerHTML += newLines;
+                } else {
+                    panel.innerHTML = t('ol')(
+                        {
+                            class: `${cssBaseClass}__log_line_container`,
+                            dataElement: LOG_LINES,
+                        },
+                        lines.map((line) => this.renderLogLine(line)).join('\n')
+                    );
+                }
+
+                // if we're autoscrolling, scroll to the bottom
+                if (this.state.scrollToEndOnNext) {
+                    this.doScrollToBottom();
+                }
+                this.renderButtons();
+            }
+
+            /**
+             * Update the log control buttons
+             * Also renders the widget state if debugging mode is on
+             */
+            renderButtons() {
```

Comment:
New -- this was incorporated into the FSM machinery, but simplifying the widget has meant it can be pulled out into its own function.